### PR TITLE
add DPO and SFT of TRL support in Gaudi and example

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
should work with https://github.com/huggingface/optimum-habana/pull/600

**Comment by sywangyi** [comment]
@regisss do we have plan to enable TRL in Gaudi? I have enabled PPO, DPO and SFT by my side. If we have the plan. I will upload the PR one by one.

**Comment by sywangyi** [comment]
add @libinta for comment

File: examples/trl_example/stack_llama_2/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# DPO pipeline for the creation of StackLlaMa 2: a Stack exchange llama-v2-7b model
```
Comment (libinta) [inline examples/trl_example/stack_llama_2/README.md:1]: can we have folder structure like
example/trl/stack_llama_2 ? try to be close to the original trl git

Comment (sywangyi) [reply examples/trl_example/stack_llama_2/README.md:1]: I just use the script under https://github.com/huggingface/trl/tree/main/examples/research_projects/stack_llama_2/scripts. what kind of folder structure are you suggesting, I will follow it.

Comment (libinta) [reply examples/trl_example/stack_llama_2/README.md:1]: I would suggest to use example/trl/stack_llama_2

File: examples/trl_example/stack_llama_2/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+```
+$ pip install -U -r requirements.txt
+```
+
+## Training
+
+There were two main steps to the DPO training process:
+1. Supervised fine-tuning of the base llama-v2-7b model to create llama-v2-7b-se:
+    - `python ../../gaudi_spawn.py --world_size 8 --use_mpi sft_llama2.py --training_args.output_dir="sft_output" --training-args.report_to none`
```
Comment (libinta) [inline examples/trl_example/stack_llama_2/README.md:15]: i saw there are many passed-in arguments from [example/trl/stack_llama_2](https://github.com/huggingface/trl/tree/main/examples/research_projects/stack_llama_2/scripts . these are not needed?

File: examples/trl_example/stack_llama_2/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+1. Run the DPO trainer using the model saved by the previous step:
```
Comment (libinta) [inline examples/trl_example/stack_llama_2/README.md:16]: should be step 2

File: examples/trl_example/stack_llama_2/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    - `python ../../gaudi_spawn.py --world_size 8 --use_mpi dpo_llama2.py --model_name_or_path="sft_output/final_merged_checkpoint" --output_dir="dpo_output" --report_to=none`
+
```
Comment (libinta) [inline examples/trl_example/stack_llama_2/README.md:18]: what about Merging the adaptors? have you tried it?

File: examples/trl_example/stack_llama_2/README.md (hunk 2 above)
Comment (sywangyi) [reply examples/trl_example/stack_llama_2/README.md:15]: I fixed it in https://github.com/huggingface/optimum-habana/pull/600, make sure your codebase contains this fix

File: examples/trl_example/stack_llama_2/README.md (hunk 4 above)
Comment (sywangyi) [reply examples/trl_example/stack_llama_2/README.md:18]: merged adapter are in SFT part. see https://github.com/huggingface/trl/blob/main/examples/research_projects/stack_llama_2/scripts/sft_llama2.py#L183C1-L187C66

File: examples/trl_example/stack_llama_2/dpo_llama2.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            name for name, buffer in model.named_buffers() if buffer.dtype == torch.bool
+        ]
+
+    model_ref = AutoModelForCausalLM.from_pretrained(
+        script_args.model_name_or_path,
+        low_cpu_mem_usage=True,
+        torch_dtype=torch.bfloat16,
+    )
+    model_ref.config.use_cache = False
+    tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply examples/trl_example/stack_llama_2/dpo_llama2.py:170]: ok, I will modify it. this part is just copying from trl example

**Review by libinta — APPROVED** [review]

**Review by libinta — COMMENTED** [review]
can you add a link in  each .py file to specify where this file is ported from and the version also?

**Comment by sywangyi** [comment]
> can you add a link in each .py file to specify where this file is ported from and the version also?

done

**Pull request merged**